### ISSUE TITLE
Modding

### ISSUE THREAD

**Issue by SnowSnakz** [SnowSnakz/fragment-faring#3]
## Compiler
[ ] Read mod.json, description.md and credits.md to create MOD header.
[ ] Convert traditional (.png, .jpg, etc.) texture files into GZ compressed raw pixel data.
[ ] Convert traditional (.ogg, .wav, .mp3) audio files into GZ compressed PCM streams.
[ ] Read JSON files and create lighter binary representations of scriptable objects.
[ ] Minify and compress referenced scripts.

## Loading
[ ] Implement mod loading on separate thread (client) to have a smoother loading screen
[ ] Read MOD header
[ ] Load textures and audio
[ ] Load scriptable objects
[ ] Assign a serial_id to all loaded scriptable objects (use saved ids for matching `mod:object_name` pairs.)

**Comment by Nosh-Ware** [comment]
Ima set out to make a simple json minifying library and get back to ya...

**Comment by Nosh-Ware** [comment]
Also I realize load time can become an issue but I feel like converting every bit of data into a binary representation may be a bit overboard. I made this commit outlining it just being json files and whatever other source files in a zip with a mustache on basically. Perhaps textures and audio can be lossfully compressed, have metadata stripped, and be preconverted into a image format that works well with the engine and maybe if you intend for the dimension of images to be always constant than maybe raw pixel data could work but like... don't over do it.

Also added zlib but if that not the best choice of compression lib then we should just stick to one and use it for all types of compression if thats possible.

**Comment by Nosh-Ware** [comment]
Also I really dont think making every json file into a binary counterpart makes sense, specifically in c where there is no trivial serialization option...

Also its just like a bit of text it need to read through we dont need to load whole dictionaries just look for the keys we expect in the json file.

**Comment by Nosh-Ware** [comment]
Maybe raw pixel data could be a option for developers when a tileset or sprite thing is found that is compatible. This format could be used alot in the base pack however, helping it load and also because we have direct control over the constraints over what that raw pixel data has. However this shouldn't be abused, it should effectively be just whatever a tile loaded from another file type would turn into in the engine.

**Comment by Nosh-Ware** [comment]
However audio data has no reason to be dealing with that. Its a few samples that would be loaded in intermitently and kept in memory. There are already perfectly good solutions to account for loading time, compression/decompression, and that fact that all audio is already in these established formats. Worst case scenario a file loads in a bit late, or if audio causes lag, there can be an option to reduce audio or turn it off. Its not like images where the game wont work at all without them, or that there would be so many on screen. Also it means that moths wont be wasted on weird proprietary formats that most likely will not help in the long run, as well as possible future porting issues.

**Comment by Nosh-Ware** [comment]
If worse come to worse for sfx then perhaps atlasing the small sound effects together would work. Again the loading is now only loading in a single type of audio file (ie .wav or .mp3 or whatever is chosen) so you simply list the timestamps and play the stream starting from there.

**Comment by Nosh-Ware** [comment]
In terms of the compiler it should also monitor for any unused files and not include them, for the sake of decompression time...

**Comment by SnowSnakz** [comment]
@Nosh-Ware I would like the solution to loading runtime assets to be as independent from 3rd party libraries as possible.
Things like the Mod.json file will be compiled into a more static format (raw cstrings, integers etc), the only reason it exists as a json file before-hand is so mod developers will have an easier time changing it.

**Comment by SnowSnakz** [comment]
As for the audio and texture stuff, I don't want to rely on third party libraries for loading those, it would be easier to include them as raw pixel/sample data so that the game doesn't have to worry about formats or anything like that, it can just run with it. Especially for textures, since I will be creating a texture packer that executes at runtime for things like tile textures, which will ideally all need to be loaded onto the same texture so that multiple types of tiles can exist in the same mesh without requiring additional shader code to select the right texture.

**Comment by SnowSnakz** [comment]
I also would like to point out that `.mod` files are only meant to be an easier format to load by the program, they aren't designed to be decompiled or edited after they are created.

**Comment by Nosh-Ware** [comment]
Fair enough but it needs to be said that this is the harder route to do this all.

May I interest you in a [kaitai](https://doc.kaitai.io/) in this trying time?
